### ISSUE TITLE
Feature request: Modify host header from client

### ISSUE THREAD

**Issue by metroid09** [inlets/inlets#82]
## Expected Behaviour

I'd like to propose an option in inlets to modify the host header from the client side of the tunnel. Something like `inlets client --remote=my.remote.live:8000 --upstream=127.0.0.1:80 --host=https://my.dev.localhost` which would change the hosts header of the packets on the client machine to `Host: my.dev.localhost` for use with internal reverse proxies.
## Current Behaviour

Current behavior leaves the host header as `Host: my.remote.live` which doesn't allow for internal reverse proxies to route the traffic properly.

## Context

I'm currently using [https://github.com/jwilder/nginx-proxy](url) to route traffic internally for testing with a multi service configuration involving several different APIs all making requests to each other. This allows me to test in a fashion which is very close to our production environment. Due to this reverse proxying, when I try to use a command like `inlets client --remote=inlet.mysite.dev:2222 --upstream=https://mysite1.dev.localhost:443` it does not connect me to the proper localhost environment `mysite1.dev.localhost` but instead connects me to the rollover site on the reverse proxy `mysite0.dev.localhost`. Having a flag for changing the host header would allow this configuration to work, and may be helpful for other use cases outside my own.

## Your Environment

* inlets version 
`Version: 2.2.0
Git Commit: 2f5e458d062e55dda9f08109f7b2c3c6919fcdf9` 

* Docker Version:`19.03.1`

* Operating System and version (e.g. Linux, Windows, MacOS): Ubuntu 18.04

**Comment by alexellis** [comment]
The `--upstream` URL can take multiple DNS names and mappings. If you create the same name or host entry locally, it should map.

I.e.

``
--upstream="my.remote.live:8000=http://my.dev.localhost:8000"
```

**Comment by metroid09** [comment]
Hi Alex,

Thanks for the response. I've attempted a couple of different iterations of the upstream URL to make this work, and so far I've been unsuccessful. I'll work on getting a docker-compose file setup with a minimum configuration that will show what I'm talking about.

**Comment by alexellis** [comment]
Please provide what you've tried.

I will take a look when I have time and see if I can help you.

**Comment by dprandzioch** [comment]
@alexellis I think I have a different problem, maybe I can give some more insights. Here's the use-case:

I develop an application with host-based multitenancy, so tenant1.local, tenant2.local etc. For purposes of testing from remote machines, I use inlets to be able to access my local development environment from the outside. When requesting the application, the requested host is resolved to a tenant and its respective database. Therefore, if I want to end up in the correct tenant, the request has to be dispatched with the proper Host header of "tenant1".

In earlier versions of inlets, when I would run `inlets client ... --upstream=http://tenant1.local`, inlets would forward all requests to `http://exit.tld/index.php` with the Host header of "tenant1.local" and therefore internally request `http://tenant1.local/index.php` instead of `http://exit.tld/index.php` which is the current behavior.

So I guess, what @metroid09 wants is the same what I would need which is to give inlets a host header, that I want it to rewrite on each request so that my application is being requested with `tenant1.local` instead of `exit.tld` which is the hostname of my exit node.

As a workaround I can configure my tenant to use `exit.tld` as a host, but it would be easier to be able to keep using `tenant1.local` and therefore have it working both locally and when requested from remote.

**Comment by alexellis** [comment]
Perhaps @ibuildthecloud may be able to make a suggestion on this? I wrote the previous tunnelling code myself but the new version has a dependency on Rancher's remotedialer package.

I'm surprised that the `--upstream=http://tenant1.local` isn't proxying with the header `Host: tenant1.local`, but I wonder what should happen in this circumstance in HTTP?

**Comment by alexellis** [comment]
Are you finding this to be an issue still?

**Comment by dstotijn** [comment]
> I'm surprised that the `--upstream=http://tenant1.local` isn't proxying with the header `Host: tenant1.local`, but I wonder what should happen in this circumstance in HTTP?

I think the behavior of inlets server and remotedialer to leave the `Host` header untouched is expected default behavior for reverse proxies. Adding a `--host-header` option to inlets to rewrite it, similar to what ngrok has, would be a nice new feature.

@metroid09 @dprandzioch I've also had the need to rewrite the `Host` header because my local app expected a specific address that didn't match the host used to access my remote/exit server. I've managed to work around it by using Caddy (with automatic HTTPS being a nice added benefit) or nginx in front of inlets on the exit server and using their header rewrite capabilities.

Example config for nginx:
```
location / {
	proxy_pass http://localhost:8090;
	proxy_set_header Host my.dev.localhost;
}
```

Example config for Caddy (v2):
```
reverse_proxy localhost:8090 {
	header_up Host my.dev.localhost
}
```

**Comment by alexellis** [comment]
Hi @dstotijn thank you for your interest.

It sounds like you may benefit from using inlets PRO, where you get a pure TCP stream, and can do whatever you like. In effect, your Caddy instance runs on your client / laptop, as pure as it gets.

Alex

**Comment by alexellis** [comment]
Here's a quick example: https://docs.inlets.dev/#/get-started/quickstart-caddy

Pros: it's all automated, no hacking on config files, no messing about on the server. Everything runs on your client so you can delete a server and recreate it at any time. The control-plane web socket is also encrypted by default.

Another example: https://blog.alexellis.io/expose-grafana-dashboards/

**Comment by alexellis** [comment]
I'm still struggling to understand the use-case here.

You have two sub-domains, both point at different local servers?

```
--upstream http://tenant1.example.com=tenant1.local:8080,http://tenant2.example.com=tenant2.local:8080
```

Why doesn't that work? The HTTP server / API on `tenant1.local:8080` will receive a HTTP request, and the domain will be as published `tenant1.example.com`

Why is that a problem? Help me understand? A full example where I can reproduce this would be most helpful.

**Comment by alexellis** [comment]
@metroid09 your use-case sounds like it would benefit from using a pure TCP proxy, so that you can run that Nginx instance locally, and not on the exit-server. See my comments above, you can try it out for free for 14 days.

**Comment by dstotijn** [comment]
> I'm still struggling to understand the use-case here.
> 
> You have two sub-domains, both point at different local servers?
> 
> ```
> --upstream http://tenant1.example.com=tenant1.local:8080,http://tenant2.example.com=tenant2.local:8080
> ```
> 
> Why doesn't that work? The HTTP server / API on `tenant1.local:8080` will receive a HTTP request, and the domain will be as published `tenant1.example.com`
> 
> Why is that a problem? Help me understand? A full example where I can reproduce this would be most helpful.

Here's what I believe the use case is of the issue author, and something I've also had to deal with occasionaly:
* HTTP service running on local computer is very picky about the `Host` HTTP header of incoming requests. E.g. it will redirect/deny any reqs that don't have `Host: my.dev.localhost`.
* The exit server is not being accessed via `my.dev.localhost` of course, but on a FQDN that's publicly resolvable, e.g. `my.remote.live`.
* Thus, the `Host` header that's tunneled all the way down to the local HTTP service will have `my.remote.live` as the value. And because the HTTP service running on the local computer is picky and expects only `my.dev.localhost`, it will cause unexpected behaviour.

Hope that makes sense! One could argue this is a bit of an edge case or something that warrants a more pure TCP proxy like you mentioned.

**Comment by dstotijn** [comment]
> Hi @dstotijn thank you for your interest.
> 
> It sounds like you may benefit from using inlets PRO, where you get a pure TCP stream, and can do whatever you like. In effect, your Caddy instance runs on your client / laptop, as pure as it gets.
> 
> Alex

My usage has been pretty minimal and sporadic so far (and non-work/team) so the OSS version of inlets manually configured on a VPS I already had has been good enough and is working really well. But I'll keep PRO in mind if I ever need more from it. 👍

**Comment by alexellis** [comment]
Thanks for ELI5 the use-case here, I think I get it now, I just don't get why the OP has that issue as I've never run into it. I also checked, and even with domain mapping in the `--upstream` flag in inlets OSS, the original name of the exit-server does get passed through.

<img width="3008" alt="Screenshot 2020-10-27 at 13 26 16" src="https://user-images.githubusercontent.com/6358735/97308602-f4e88580-1858-11eb-926c-7d67fe11ac42.png">

@dstotijn I hear you, however perhaps you'd be open to just kicking the tires with PRO? You'll also get access to the new project we're working on "inlets plus" = "inlets OSS plus more features" - the trial is free, and you sound like you'd be able to put it through its paces. There is a special discounted price for personal use available.

**Comment by dstotijn** [comment]
> I also checked, and even with domain mapping in the --upstream flag in inlets OSS, the original name of the exit-server does get passed through.

Indeed. The `Host` header header getting passed through as-is (e.g. `www.example.com` in your screenshot), which is expected and desired for most users. Typically a local web service/upstream doesn't have any logic dependent on a specific `Host` HTTP header and there is no issue; that's likely why you've never run into it.

But if the local app is picky about the `Host` or you have some internal proxy/server infra (the issue author mentions he's locally running [nginx-proxy](https://github.com/jwilder/nginx-proxy), the ngrok docs [mention](https://ngrok.com/docs#http-host-header) WAMP/MAMP as an example) that's configured to work with hardcoded/specific hosts, it becomes necessary to rewrite the `Host` header.

**Comment by alexellis** [comment]
I would suggest two paths forward for the feature request.

1) use [inlets PRO](https://inlets.dev/) and run Nginx or Caddy close to where the traffic is (on the internal server) instead of on the exit-server, which is meant to be stateless. In this way your exit-server needs no maintenance, and can be deleted / re-created at any time with no configuration on the host.
2) request this feature as a flag for [inlets plus](https://github.com/inlets/inlets-plus) - this will be inlets OSS, but with some additional features on the roadmap like HTTPS certs from LetsEncrypt, local tunnelling (as well as remote), mTLS for the control-plane and more.

Thanks for your input @dstotijn that was very helpful.
